ingestion-beam/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BatchLoads.java

```diff
@@ -0,0 +1,638 @@
+import org.apache.beam.sdk.transforms.windowing.AfterFirst;
+import org.apache.beam.sdk.transforms.windowing.AfterPane;
+import org.apache.beam.sdk.transforms.windowing.AfterProcessingTime;
+import org.apache.beam.sdk.transforms.windowing.DefaultTrigger;
+import org.apache.beam.sdk.transforms.windowing.GlobalWindows;
+import org.apache.beam.sdk.transforms.windowing.Repeatedly;
+import org.apache.beam.sdk.transforms.windowing.Window;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionList;
+import org.apache.beam.sdk.values.PCollectionTuple;
+import org.apache.beam.sdk.values.PCollectionView;
+import org.apache.beam.sdk.values.ShardedKey;
+import org.apache.beam.sdk.values.TupleTag;
+import org.apache.beam.sdk.values.TupleTagList;
+import org.apache.beam.sdk.values.TypeDescriptor;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.annotations.VisibleForTesting;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Strings;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.Lists;
+import org.joda.time.Duration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** PTransform that uses BigQuery batch-load jobs to write a PCollection to BigQuery. */
+class BatchLoads<DestinationT, ElementT>
+    extends PTransform<PCollection<KV<DestinationT, ElementT>>, WriteResult> {
+  static final Logger LOG = LoggerFactory.getLogger(BatchLoads.class);
+
+  // The maximum number of file writers to keep open in a single bundle at a time, since file
+  // writers default to 64mb buffers. This comes into play when writing dynamic table destinations.
+  // The first 20 tables from a single BatchLoads transform will write files inline in the
+  // transform. Anything beyond that might be shuffled.  Users using this transform directly who
+  // know that they are running on workers with sufficient memory can increase this by calling
+  // BatchLoads#setMaxNumWritersPerBundle. This allows the workers to do more work in memory, and
+  // save on the cost of shuffling some of this data.
+  // Keep in mind that specific runners may decide to run multiple bundles in parallel, based on
+  // their own policy.
+  @VisibleForTesting static final int DEFAULT_MAX_NUM_WRITERS_PER_BUNDLE = 20;
+
+  @VisibleForTesting
+  // Maximum number of files in a single partition.
+  static final int DEFAULT_MAX_FILES_PER_PARTITION = 10000;
+
+  @VisibleForTesting
+  // TODO: Remove this custom lower threshold; this forces a day of main pings
+  // to be split to about 100 pieces, sidestepping the bug where a single large
+  // BQ load into a clustered tables often fails with memory exceeded errors.
+  static final long DEFAULT_MAX_BYTES_PER_PARTITION = 100 * (1L << 30);
+  // // Maximum number of bytes in a single partition -- 11 TiB just under BQ's 12 TiB limit.
+  // static final long DEFAULT_MAX_BYTES_PER_PARTITION = 11 * (1L << 40);
```

Comment:
This is the one change in this file. Otherwise, it is copied verbatim from Beam 2.16.